File: bench/py/46_matrix_mult.py
# naive O(n^3) matrix multiply over lists-of-lists (nested subscripting)
import sys
scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1

def mk(n, seed):
    m = [None] * n
    x = seed
    for i in range(n):
        row = [0] * n
        for j in range(n):
            x = (x * 1103515245 + 12345) % 1000
            row[j] = x
        m[i] = row
    return m

def matmul(a, b, n):
    c = [None] * n
    for i in range(n):
        row = [0] * n
        for j in range(n):
            s = 0
            for k in range(n):
                s += a[i][k] * b[k][j]
            row[j] = s
        c[i] = row
    return c

n = 70
a = mk(n, 1)
b = mk(n, 2)

checksum = 0
for r in range(scale):
    c = matmul(a, b, n)
    checksum += c[0][0] + c[n - 1][n - 1]

print("result:", checksum)

File: bench/cpp/46_matrix_mult.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/46_matrix_mult.my: naive O(n^3) matmul over
 * arrays-of-arrays (nested subscripting). vector<vector<long>> mirrors the
 * boxed 2-D structure; the LCG fill matches the .my so the checksum agrees. */
#include "bench.h"

typedef std::vector<std::vector<long>> Mat;

static Mat mk(int n, long seed)
{
    Mat m(n);
    long x = seed;
    for (int i = 0; i < n; i++) {
        std::vector<long> row(n);
        for (int j = 0; j < n; j++) {
            x = (x * 1103515245 + 12345) % 1000;
            row[j] = x;
        }
        m[i] = std::move(row);
    }
    return m;
}

static Mat matmul(const Mat &a, const Mat &b, int n)
{
    Mat c(n);
    for (int i = 0; i < n; i++) {
        std::vector<long> row(n);
        for (int j = 0; j < n; j++) {
            long s = 0;
            for (int k = 0; k < n; k++)
                s += a[i][k] * b[k][j];
            row[j] = s;
        }
        /* the checksum only reads c[0][0] + c[n-1][n-1], so without this the
         * compiler DCEs the other O(n^2) elements (computing 2, not the full
         * matrix). MyLang computes the WHOLE matrix - force C++ to as well. */
        bench_sink_ptr(row.data());
        c[i] = std::move(row);
    }
    return c;
}

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);
    const int n = 70;
    Mat a = mk(n, 1), b = mk(n, 2);

    long checksum = 0;
    for (long r = 0; r < scale; r++) {
        Mat c = matmul(a, b, n);
        checksum += c[0][0] + c[n - 1][n - 1];
    }
    printf("result: %ld\n", checksum);
    return 0;
}
